### PR TITLE
[ADD] hr_cae_contract: termination contract, amendments view, data from employee

### PR DESCRIPTION
**Pull request by mclaeysb**
- This adds way to terminate contracts: next to ammendments, a 'termination contract' can be added. When it is put in state 'running', all other contracts in the contract group are terminated.
- The contract group view is expanded to easily see how wage evolves over contracts
- Some data is ported from the employee when a contract is made.
- Some fields were renamed and some small UI enhancements are added.

Question: I initially programmed the first functionallity listed above with an onchange but 1) this means it only works when the state is changed in that contract's form view (not when in a kanban view or when through a database write) and 2) I don't know how to test onchange. So I did it with a check on the write function (old onchange still commented out below). Would like your opions!

### CODE REVIEW

File: hr_cae_contract/data/hr_contract_data.xml [hunk 1]
```diff
@@ -22,8 +22,8 @@
         </record>
 
         <!--Accompagnement-->
-        <record id="hr_contract_type_coaching" model="hr.contract.type">
-            <field name="name">Coaching</field>
+        <record id="hr_contract_type_accompaniment" model="hr.contract.type">
+            <field name="name">Accompaniment</field>
```
Comment (robinkeunen) [inline hr_cae_contract/data/hr_contract_data.xml]: I don't think that [accompaniment](https://www.merriam-webster.com/dictionary/accompaniment) is the right term, what are you trying to say?

File: hr_cae_contract/models/hr_contract.py [hunk 2]
```diff
@@ -257,6 +308,12 @@ def _compute_latest(self):
                 contract.contract_group_id.get_latest()
             )
 
+    def _compute_has_termination_contract(self):
+        for contract in self:
+            contract.has_termination_contract = (
+                contract.contract_group_id.has_termination_contract()
+            )
```
Comment (robinkeunen) [inline hr_cae_contract/models/hr_contract.py:313]: I think this layer is not needed, you could do `has_termination_contract()` computation directly here.

File: hr_cae_contract/models/hr_contract.py [hunk 3]
```diff
@@ -50,15 +50,28 @@ def get_latest(self):
...
 
     echelon = fields.Selection(
-        [("main", "Main"), ("amendment", "Amendment")],
+        [
+            ("main", "Main"),
+            ("amendment", "Amendment"),
+            ("termination", "Termination"),
+        ],
         string="Echelon",
-        default="main",
```
Comment (robinkeunen) [inline hr_cae_contract/models/hr_contract.py:61]: Why did you remove the default value?

File: hr_cae_contract/models/hr_contract.py [hunk 4]
```diff
@@ -73,21 +86,42 @@ class ContractTag(models.Model):
...
+            ("close", "Expired"),
+            ("termination", "Terminated"),
+            ("cancel", "Cancelled"),
+        ],
+        copy=False,
+    )
     employee_id = fields.Many2one(required=True)
-    type_echelon = fields.Selection(
-        [("main", "Main"), ("amendment", "Amendment")],
+    echelon = fields.Selection(
```
Comment (robinkeunen) [inline hr_cae_contract/models/hr_contract.py:108]: 👍

File: hr_cae_contract/models/hr_contract.py [hunk 5]
```diff
@@ -73,21 +86,42 @@ class ContractTag(models.Model):
 class Contract(models.Model):
     _inherit = "hr.contract"
 
+    @api.model
+    def _default_type_id(self):
```
Comment (robinkeunen) [inline hr_cae_contract/models/hr_contract.py]: function name should start by a verb.

File: hr_cae_contract/models/hr_contract.py [hunk 6]
```diff
@@ -185,6 +224,26 @@ def create(self, vals):
             res._compute_initial()
         return res
 
+    # Alternative for onchange (see below) such that it works on every write
+    @api.multi
+    def write(self, vals):
+        res = super(Contract, self).write(vals)
+        for contract in self:
+            if (
+                bool(vals.get("state"))
```
Comment (robinkeunen) [inline hr_cae_contract/models/hr_contract.py:231]: Why not a `@api.depends`?

File: hr_cae_contract/models/hr_contract.py [hunk 7]
```diff
@@ -185,6 +224,26 @@ def create(self, vals):
...
+                and contract.echelon == "termination"
+                and contract.state == "open"
+            ):
+                other_contracts = contract.contract_ids - contract
```
Comment (robinkeunen) [inline hr_cae_contract/models/hr_contract.py:235]: clean.

File: hr_cae_contract/models/hr_contract.py [hunk 8]
```diff
@@ -321,30 +385,47 @@ def check_type_count(self):
...
+        self.ensure_one()
+        if self.has_termination_contract:
+            raise ValidationError(
+                _("This contract group already has a termination contract ")
+            )
         if not type_id:
             type_id = self.env["hr.contract.type"].search(
-                [("echelon", "=", "amendment")], limit=1
+                [("echelon", "=", echelon)], limit=1
             )
```
Comment (robinkeunen) [inline hr_cae_contract/models/hr_contract.py:402]: Explicitly look for main?

File: hr_cae_contract/models/hr_contract.py (hunk 2 above)
Comment (robinkeunen) [reply hr_cae_contract/models/hr_contract.py:313]: You should create a function if

You use the same code in several places,
You need to break up a long process in smaller units for better readability.
... ?

File: hr_cae_contract/models/hr_contract.py [hunk 9]
```diff
@@ -321,30 +385,47 @@ def check_type_count(self):
...
                 "state": "draft",
+                "echelon": echelon,
                 "type_id": type_id.id,
                 "duration": False,
                 "date_end": False,
                 "amendment_index": latest_contract_id.amendment_index + 1,
             }
         )
-        amendment.check_type_count()
+        contract.check_type_count()
```
Comment (robinkeunen) [inline hr_cae_contract/models/hr_contract.py:420]: Why do you need to  launch manually this constraint?

**Review by robinkeunen — CHANGES_REQUESTED** [review]
As said in the comment: why not use `@api.depends` instead of `@api.onchange` or overiding the write method?

To test an onchange methode, you need to use the Form class :-)

File: hr_cae_contract/models/hr_contract.py (hunk 3 above)
Comment (mclaeysb) [reply hr_cae_contract/models/hr_contract.py:61]: Because otherwise it sets this as default for the echelon of the three existing contract types in Odoo HR, and these three also start showing up when making a new main type contract.

File: hr_cae_contract/data/hr_contract_data.xml (hunk 1 above)
Comment (mclaeysb) [reply hr_cae_contract/data/hr_contract_data.xml]: Indeed, a CAE is no orchestra. Accompaniment > Support ✔️ 🙏

File: hr_cae_contract/models/hr_contract.py (hunk 2 above)
Comment (mclaeysb) [reply hr_cae_contract/models/hr_contract.py:313]: I'm not calling `contract.has_termination_contract()` but `contract.contract_group_id.has_termination_contract()`. Searching through existing Odoo code, I don't think thats possible `_compute` requires a string), but I might be wrong?

File: hr_cae_contract/models/hr_contract.py (hunk 3 above)
Comment (mclaeysb) [reply hr_cae_contract/models/hr_contract.py:61]: Another option might be to launch a pre-hook and give all existing types a echelon e.g. `other`, but that seems a bit overkill?
The default was not really useful anyway.

File: hr_cae_contract/models/hr_contract.py (hunk 5 above)
Comment (mclaeysb) [reply hr_cae_contract/models/hr_contract.py]: Thanks! 
-> `_get_default_type_id`

File: hr_cae_contract/models/hr_contract.py (hunk 8 above)
Comment (mclaeysb) [reply hr_cae_contract/models/hr_contract.py:402]: Don't know what you mean here..

File: hr_cae_contract/models/hr_contract.py (hunk 9 above)
Comment (mclaeysb) [reply hr_cae_contract/models/hr_contract.py:420]: Because even though `type_id` is set on this copy and is also mentioned in the @api.constrains, it wasn't launched manually. I was puzzeled as well. Maybe I'm missing something? Or it's cleaned to overwrite the `copy`? Any Idea? Google doesn't seem to know.
The [documentation](https://www.odoo.com/documentation/12.0/reference/orm.html) (> `contrains`) doesn't mentions notes about `write` and `create` but doesn't mention anything about `copy`...

File: hr_cae_contract/models/hr_contract.py (hunk 6 above)
Comment (mclaeysb) [reply hr_cae_contract/models/hr_contract.py:231]: Had looked this up in the existing Odoo code. I have only seen `@api.depends` used in computed fields (which is not the case the case here since it should still be clickable). Would be glad to be pointed to some code if you know of `@api.depends` being used in such a context!

**Comment by mclaeysb** [comment]
Thanks for the review @robinkeunen ! Answered to your questions. Let me know what you think!
Really puzzeled still that an `api.onchange` can't be used in what seems to be it's main use-case: do something when a value changes.

File: hr_cae_contract/models/hr_contract.py (hunk 3 above)
Comment (robinkeunen) [reply hr_cae_contract/models/hr_contract.py:61]: good reasoning, ok to remove the default value 👍

File: hr_cae_contract/models/hr_contract.py (hunk 6 above)
Comment (robinkeunen) [reply hr_cae_contract/models/hr_contract.py:231]: If you don't have a link to `other_contracts` on this contract then, indeed, this is a good way to do it :-)

File: hr_cae_contract/models/hr_contract.py (hunk 8 above)
Comment (robinkeunen) [reply hr_cae_contract/models/hr_contract.py:402]: I don't now either, nevermind 😅

File: hr_cae_contract/models/hr_contract.py (hunk 9 above)
Comment (robinkeunen) [reply hr_cae_contract/models/hr_contract.py:420]: Mmh, it's probably good to trigger it here then, good catch.

File: hr_cae_contract/models/hr_contract.py (hunk 2 above)
Comment (robinkeunen) [reply hr_cae_contract/models/hr_contract.py:313]: Ok, I read too fast it seems :-)

**Review by robinkeunen — APPROVED** [review]
Good call on everything I raise :-) C'est du travail rigoureux!

**Comment by mclaeysb** [comment]
@legalsylvain Am I doing something wrong?

**Comment by legalsylvain** [comment]
No. All is good. My ocabot is just sleeping right now, because i'm testing a new feature. I'll ping you when he's awake to test again. 
Regards.

**Comment by legalsylvain** [comment]
Hi. @mclaeysb. my bot is awake. You can test again !

Nitpicking, 
However, not sure the option ``patch`` is correct, regarding the diff. Doesn't seem to be a patch, or a bug fix. Don't you think ? I guess a ``minor`` option is better.

Regarding the [OCA convention](https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#version-numbers)

thanks !

**Comment by legalsylvain** [comment]
Nice !
